core/control-plane/contract/src/main/java/org/eclipse/dataspaceconnector/contract/offer/ContractOfferServiceImpl.java

```diff
@@ -60,10 +63,22 @@ public Stream<ContractOffer> queryContractOffers(ContractOfferQuery query) {
         var agent = agentService.createFor(query.getClaimToken());
 
         return definitionService.definitionsFor(agent, query.getRange())
+                // ignore definitions with missing criteria (only if not SELECT_ALL)
+                .filter(definition -> (definition.getSelectorExpression().getCriteria() != null &&
+                        !definition.getSelectorExpression().getCriteria().isEmpty()) ||
+                        definition.getSelectorExpression() == AssetSelectorExpression.SELECT_ALL)
                 .flatMap(definition -> {
-                    var assetFilterQuery = QuerySpec.Builder.newInstance()
-                            .filter(concat(definition.getSelectorExpression().getCriteria().stream(), query.getAssetsCriteria().stream()).collect(Collectors.toList())).build();
-                    var assets = assetIndex.queryAssets(assetFilterQuery);
+                    Stream<Asset> assets;
+                    if (definition.getSelectorExpression() == AssetSelectorExpression.SELECT_ALL) {
```

Comment:
Pay attention on using `==` to equal on object instances (likely it will never be true).

Reply:
resolved